src/dlls/mscoree/mscoree.cpp

```diff
@@ -1263,17 +1233,31 @@ HRESULT SetInternalSystemDirectory()
         DWORD len;
 
         // use local buffer for thread safety
-        WCHAR wzSystemDirectory[COUNTOF(g_pSystemDirectory)];
+        WCHAR * wzSystemDirectory = new (nothrow) WCHAR[MAX_LONGPATH+1];
+        if (wzSystemDirectory == NULL)
+        {
+            return HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_MEMORY);
+        }
         
-        hr = GetCORSystemDirectoryInternal(wzSystemDirectory, COUNTOF(wzSystemDirectory), &len);
+        hr = GetCORSystemDirectoryInternal(wzSystemDirectory, MAX_LONGPATH+1, &len);
 
         if(FAILED(hr)) {
-            wzSystemDirectory[0] = W('\0');
+            *wzSystemDirectory = W('\0');
             len = 1;
         }
-
+        
+        WCHAR * pSystemDirectory = new (nothrow) WCHAR[len];
+        if (pSystemDirectory == NULL)
+        {
+            return HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_MEMORY);
```

Comment:
`wzSystemDirectory` will leak if we exit here. Could you please change the line that allocates it to:

`
NewArrayHolder<WCHAR>  wzSystemDirectory(new (nothrow) WCHAR[MAX_LONGPATH+1]);
`

and delete the line that deletes it below. NewArrayHolder will take care of deleting it no matter which path is taken to exit the function.